### PR TITLE
Feature/signin signup

### PR DESCRIPTION
**Pull request by vc-jinal**
SignUp and SignIn using JWT token

### CODE REVIEW

**Review by karansoni-vc — CHANGES_REQUESTED** [review]

File: signup-signin-jwt/index.js [hunk 1]
```diff
@@ -38,26 +38,23 @@ app.post('/login', async (req, res) => {
...
         if (!passValid) {
-            return res.json({ error: "Invalid email" })
+            return res.status(400).json({ error: "Invalid password" })
         }
         const token = jwt.sign({ email: user.email },
-            jwtSecretkey,
-            { expiresIn: '4h' });
-        return res.status(202).json({ token });
+            jwtSecretkey);
+        return res.status(202).json({ token });//acept for proessing but not completed
```
Comment (vc-mohit) [inline signup-signin-jwt/index.js:50]: @vc-jinal: use spell checker.

File: signup-signin-jwt/index.js [hunk 2]
```diff
@@ -38,26 +38,23 @@ app.post('/login', async (req, res) => {
...
     }
     catch (error) {
         return res.json({ error: "Coudln't login " })
```
Comment (vc-mohit) [inline signup-signin-jwt/index.js:53]: check message

File: signup-signin-jwt/index.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+//middleware for jwt verification
+
+const verifyToken = (req, res, next) => {
+    const token = req.headers.authorization;
+    if (!token) {
+        return res.json({ error: "Token is not given" })
+    }
+    jwt.verify(token, jwtSecretkey)
+    {
+        req.user = decoded;
```
Comment (vc-mohit) [inline signup-signin-jwt/index.js:67]: decoded is not defined

File: signup-signin-jwt/index.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+if (!fs.existsSync(userFile)) {
+    fs.writeFileSync(userFile, []);
+}
+
+app.post('/register', async (req, res) => {
+    const users = JSON.parse(fs.readFileSync(userFile));
+    const { name, email, password } = req.body;
+    const userExist = users.find((user) => user.email === email);
+    if (userExist) {
+        return res.json({ error: "user already exist.." });
```
[the remark this answers is not in the document]
Comment (vc-jinal) [reply signup-signin-jwt/index.js]: Added status code and message

File: signup-signin-jwt/index.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+    try {
+        const hashPassword = await bycrypt.hash(password, 10);
+        const user = { name, email, password: hashPassword }
+        users.push(user);
+        fs.writeFileSync(userFile, JSON.stringify(users, null, 2));
+        return res.json({ message: "User registered successfully" })
+    } catch (err) {
+        return res.json({ err: "registration failed" });
```
[the remark this answers is not in the document]
Comment (vc-jinal) [reply signup-signin-jwt/index.js]: removed sync method and added status code

File: signup-signin-jwt/index.js [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    }
+})
+
+
+app.post('/login', async (req, res) => {
+    const users = JSON.parse(fs.readFileSync(userFile, 'utf-8'));
+    const { email, password } = req.body;
+    const user = users.find((user) => user.email === email);
+    if (!user) {
+        return res.json({ error: "Invalid email or password" })
```
[the remark this answers is not in the document]
Comment (vc-jinal) [reply signup-signin-jwt/index.js]: Add message

File: signup-signin-jwt/index.js [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+
+//middleware for jwt verification
+
+const verifyToken = (req, res, next) => {
+    const token = req.headers.authorization;
+    if (!token) {
+        return res.json({ error: "Token is not given" })
+    }
+    jwt.verify(token, jwtSecretkey, (err, decoded) => {
```
[the remark this answers is not in the document]
Comment (vc-jinal) [reply signup-signin-jwt/index.js]: removed callback

File: signup-signin-jwt/index.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    const users = JSON.parse(fs.readFileSync(userFile, 'utf-8'));
+    const { email, password } = req.body;
+    const user = users.find((user) => user.email === email);
+    if (!user) {
+        return res.json({ error: "Invalid email or password" })
+    }
+    try {
+        const passValid = await bycrypt.compare(password, user.password);
+        if (!passValid) {
+            return res.json({ error: "Invalid email" })
```
[the remark this answers is not in the document]
Comment (vc-jinal) [reply signup-signin-jwt/index.js]: Add message

File: signup-signin-jwt/index.js (hunk 3 above)
Comment (vc-jinal) [reply signup-signin-jwt/index.js:71]: removed callback and defined decoded

File: signup-signin-jwt/index.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        return res.status(400).json({ message: "Invalid email" })
+    }
+    try {
+        const passValid = await bycrypt.compare(password, user.password);
+        if (!passValid) {
+            return res.status(400).json({ error: "Invalid password" })
+        }
+        const token = jwt.sign({ email: user.email },
+            jwtSecretkey);
+        return res.status(200).json({ token });
```
[the remark this answers is not in the document]
Comment (vc-jinal) [reply signup-signin-jwt/index.js]: Add message

File: signup-signin-jwt/index.js [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+const bycrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+const fs = require('fs/promises');
+require('dotenv').config();
+const userFile = process.env.MY_FILE;
+const jwtSecretkey = process.env.JWT_SECRET_KEY;
+app.use(bodyParser.json());
+
+
+if (!fs.existsSync(userFile)) {
```
Comment (vc-mohit) [inline signup-signin-jwt/index.js:13]: this function is working?

**Comment by vc-mohit** [comment]
I am closing the Pull Request because of the following reason:
1. Lots of Spelling mistakes
2. All responses should be in JSON format only. (Use Send method)
3. Code should be properly formatted.

**Pull request closed without merging**